src/library/escape.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * escape.c - Source file for escaping capability
+ * Copyright (c) 2021 Red Hat Inc.
+ * All Rights Reserved.
+ *
+ * This software may be freely redistributed and/or modified under the
+ * terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2, or (at your option) any
+ * later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING. If not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor
+ * Boston, MA 02110-1335, USA.
+ *
+ * Authors:
+ *   Radovan Sroka <rsroka@redhat.com>
+ */
+
+#include "escape.h"
+
+#include <stdio.h>
+#include <ctype.h>
+#include <string.h>
+
+#include "message.h"
+
+static int isHEX(int x)
+{
+  if (isxdigit(x) > 0 && !(islower(x) > 0))
+    return 1;
+
+  return 0;
+}
+
+static char asciHex2Bits(char X)
+{
+  char base = 0;
+  if (X >= '0' && X <= '9') {
+    base = '0';
+  } else if (X >= 'A' && X <= 'F') {
+    base = 'A' - 10;
+  }
+  return (X - base) & 0X00FF;
+}
+
+
+/*
+ * the function returns escaped copy of the input string
+ * returned string should be freed by caller
+ * it returns NULL in case of error
+ *
+*/
+char * escape(const char* input, int mode)
+{
+  char buffer[4096+1] = {0};
+  size_t pos = 0;
+
+  for (size_t i = 0 ; i < strlen(input); i++) {
+    int should_escape = 0;
+
+    switch(mode) {
+
+    case EVERYTHING:
+      should_escape = 1;
+      break;
+
+    case WHITESPACES:
+      if (isblank(input[i]))
+        should_escape = 1;
+      break;
+
+    default:
+      return NULL;
+      break;
+
+    }
+
+    // always ascape % sign !!!
+    if (input[i] == '%')
+      should_escape = 1;
+
+    if (should_escape) {
+      char buff[4] = {0}; // "%A5\0"
+      int res = snprintf(buff, 4, "%c%02X", '%', input[i]);
+
+      if (res == -1)
+        return NULL;
+
+      /*
+      buffer[pos++] = buff[0]; // '%'
+      buffer[pos++] = buff[1]; // 'A'
+      buffer[pos++] = buff[2]; // '5'
+      */
+
+      for (size_t ii = 0; ii < 3; ii++) {
+
+        if (pos >=(sizeof(buffer) - 1))
+          return NULL;
+
+        buffer[pos++] = buff[ii];
+      }
+
+    } else {
+
+      if (pos >=(sizeof(buffer) - 1))
+          return NULL;
+
+      buffer[pos++] = input[i];
+    }
+  }
+
+  return strdup(buffer);
+}
+
+char * unescape(const char * input)
```

Comment:
use the same format as in the rest of the code
not this:
char * unescape(const char * input)
but this:
char *unescape(const char *input)